### PR TITLE
Code blocks

### PR DESCRIPTION
**Pull request by kazcw**
### Pull Request Description

Implement [code blocks](https://github.com/enso-org/design/blob/wip/wd/enso-spec/epics/enso-spec-1.0/03.%20Code%20format%20and%20layout.md#code-blocks).

See: https://www.pivotaltracker.com/story/show/182496883

### Important Notes

- ~~This isn't the best solution for `resolver::run`. I'm working on a cleaner approach without recursion. However I just started feeling sick, and all the rest of the PR is ready for review--so to reduce latency I am posting this for review as is; the refactor shouldn't affect anything outside `resolver.rs` and possibly the call site of `resolver::run`.~~

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH
        `./run ide dist` and `./run ide watch`.

[ci no changelog needed]

### CODE REVIEW

File: lib/rust/metamodel/lexpr/src/lib.rs [hunk 1]
```diff
@@ -69,20 +70,27 @@ pub struct ToSExpr<'g> {
     graph:   &'g TypeGraph,
     #[derivative(Debug = "ignore")]
     mappers: BTreeMap<TypeId, Box<dyn Fn(Value) -> Value>>,
+    skip:    BTreeSet<TypeId>,
```
Comment (wdanilo) [inline lib/rust/metamodel/lexpr/src/lib.rs:73]: By reading the implementation and docs, I had hard time understanding how skip should be used. Is this described somewhere in the docs?

File: lib/rust/parser/src/lexer.rs [hunk 2]
```diff
@@ -78,10 +78,11 @@ pattern_impl_for_char_slice!(1, 2, 3, 4, 5, 6, 7, 8, 9, 10);
...
     #[deref_mut]
-    pub state:    LexerState,
-    pub input:    &'s str,
-    pub iterator: str::CharIndices<'s>,
-    pub output:   Vec<Token<'s>>,
+    pub state:            LexerState,
+    pub input:            &'s str,
+    pub iterator:         str::CharIndices<'s>,
+    pub output:           Vec<Token<'s>>,
+    pub temporary_tokens: EmptyVec<Token<'s>>,
```
Comment (wdanilo) [inline lib/rust/parser/src/lexer.rs]: There is no explanation of what is this. The name "temporary tokens" does not semantically explains what it is for and why its needed.

File: lib/rust/parser/src/lexer.rs [hunk 3]
```diff
@@ -101,9 +102,10 @@ impl<'s> Lexer<'s> {
     pub fn new(input: &'s str) -> Self {
         let iterator = input.char_indices();
         let capacity = input.len() / AVERAGE_TOKEN_LEN;
-        let output = Vec::with_capacity(capacity * mem::size_of::<Token<'s>>());
+        let output = Vec::with_capacity(capacity);
```
Comment (wdanilo) [inline lib/rust/parser/src/lexer.rs:106]: Ouch, I was sure it was fixed already :O

File: lib/rust/prelude/src/data/empty_vec.rs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+use derivative::Derivative;
+
+
+
+// ================
+// === EmptyVec ===
+// ================
+
+/// A [`std::vec::Vec`] that is constrained to be empty.
```
Comment (wdanilo) [inline lib/rust/prelude/src/data/empty_vec.rs]: Please add docs explaining an example use case of this struct.

File: lib/rust/parser/src/lexer.rs [hunk 5]
```diff
@@ -677,35 +679,32 @@ impl<'s> Lexer<'s> {
...
-                    }
-                }
             }
-            for newline in newlines {
-                self.submit_token(newline);
+            self.submit_token(token.with_variant(token::Variant::newline()));
+            line_breaks.drain(..).for_each(|token| self.submit_token(token));
+            while block_indent < self.current_block_indent {
+                let err = "Lexer internal error. Inconsistent code block hierarchy.";
+                let parent_block_indent = self.end_block().expect(err);
```
Comment (wdanilo) [inline lib/rust/parser/src/lexer.rs]: The previous code was not panicking. As I was explaining earlier - EVERY code should always lex correctly. No panic is allowed. We can attach warnings, but never panic. If there is inconsistency in the code, like:

```
foo
    bar
  baz
    bax
```

then all 3 lines should be one block, and the `baz` line should emit warning. It is also described in the language spec.

File: lib/rust/parser/src/lexer.rs [hunk 6]
```diff
@@ -876,6 +875,13 @@ mod tests {
...
                 block_end_("", ""),
             ]),
+            ("foo\n    +", vec![
+                ident_("", "foo"),
+                block_start_("", ""),
+                newline_("", "\n"),
+                operator_("    ", "+"),
+                block_end_("", ""),
+            ]),
         ]);
```
Comment (wdanilo) [inline lib/rust/parser/src/lexer.rs]: Please add tests for lexing of blocks with inconsistent indentation.

File: lib/rust/parser/src/lib.rs [hunk 7]
```diff
@@ -146,16 +147,13 @@ impl Parser {
...
-            let resolver = macros::resolver::Resolver::new_root();
-            let tree = resolver.run(&self.macros, &mut tokens);
-            let tree = expression_to_statement(tree);
-            statements.push(tree);
-        }
-        syntax::Tree::block(statements)
+        let mut tokens = tokens.into_iter();
+        let resolver = macros::resolver::Resolver::new_root();
+        let lines = resolver.run(&self.macros, &mut tokens);
+        assert_eq!(tokens.next(), None);
```
Comment (wdanilo) [inline lib/rust/parser/src/lib.rs]: This fails. Parser should never fail. These tokens can be marked as not parsed and stored in AST instead.

File: lib/rust/parser/src/lib.rs [hunk 8]
```diff
@@ -201,7 +199,7 @@ fn expression_to_binding<'a>(app: &syntax::tree::OprApp<'a>) -> Option<syntax::T
                 args.push(arg.clone());
             }
             args.reverse();
-            if let Some(rhs) = rhs && args.is_empty() {
+            if let Some(rhs) = rhs && args.is_empty() && !matches!(&*rhs.variant, Variant::BodyBlock { .. }) {
```
Comment (wdanilo) [inline lib/rust/parser/src/lib.rs]: Please refactor this check to a well named variable - the name should semantically express what it means when it is true.

File: lib/rust/parser/src/lib.rs [hunk 9]
```diff
@@ -219,6 +217,7 @@ fn expression_to_binding<'a>(app: &syntax::tree::OprApp<'a>) -> Option<syntax::T
 // === Tests ===
 // =============
 
+/*
```
Comment (wdanilo) [inline lib/rust/parser/src/lib.rs]: commented code.

File: lib/rust/parser/src/macros/resolver.rs [hunk 10]
```diff
@@ -209,12 +209,13 @@ impl<'s> Resolver<'s> {
         reserved.and_option_from(|| self.macro_stack.pop())
     }
 
-    /// Run the resolver. Returns the resolved AST.
+    /// Run the resolver. Returns the resolved AST for the top-level lines of the input.
```
Comment (wdanilo) [inline lib/rust/parser/src/macros/resolver.rs]: I dont understand what does "for the top-level lines of the input" mean.

File: lib/rust/parser/src/macros/resolver.rs [hunk 11]
```diff
@@ -232,9 +233,38 @@ impl<'s> Resolver<'s> {
...
         next_token!();
-        while let Some(token) = opt_item && !token.is_newline() {
+        let mut lines = vec![];
+        let mut line_contains_items = false;
+        while let Some(token) = opt_item {
+            if let syntax::Item::Token(Token {
+                variant: token::Variant::Newline(_),
+                left_offset,
+                code,
+            }) = &token
```
Comment (wdanilo) [inline lib/rust/parser/src/macros/resolver.rs]: Ugh, this is so ugly. The worst thing is that I dont see how to make it prettier in Rust -.-

File: lib/rust/parser/src/macros/resolver.rs [hunk 12]
```diff
@@ -232,9 +233,38 @@ impl<'s> Resolver<'s> {
...
+            {
+                let newline = token::newline(left_offset.clone(), code.clone());
```
Comment (wdanilo) [inline lib/rust/parser/src/macros/resolver.rs]: is it necessary to use clones here?

File: lib/rust/parser/src/syntax/item.rs [hunk 13]
```diff
@@ -73,6 +83,30 @@ impl<'s> TryAsRef<Item<'s>> for Item<'s> {
     }
 }
 
+/// Build an AST block from the contents of an `Item::Block`.
```
Comment (wdanilo) [inline lib/rust/parser/src/syntax/item.rs]: I don't see the relation between this doc and the function signature. It tells that it builds a block out of contents of an `Item::Block`, but its signature does not accept `Item::Block`. It seems you can pass anything to it that is an iterator. Also, what does it mean that it "builds" a block? Its implementation is pretty complex and it does not explain in the doc what t really does.

File: lib/rust/parser/src/syntax/operator.rs [hunk 14]
```diff
@@ -45,77 +45,80 @@ impl<T> WithPrecedence<T> {
...
-/// It is extended to handle operator sections.
-#[inline(always)]
+/// Take [`Item`] stream, resolve operator precedence and return the final AST.
+///
+/// The precedence resolution algorithm is based on the Shunting yard algorithm[1], extended to
+/// handle operator sections.
+/// [1]: https://en.wikipedia.org/wiki/Shunting_yard_algorithm
 pub fn resolve_operator_precedence<'s>(items: Vec<syntax::Item<'s>>) -> syntax::Tree<'s> {
+    // If `items` is empty, opt_rhs.unwrap() will panic below. Let's panic now with a clear error.
+    assert!(!items.is_empty());
```
Comment (wdanilo) [inline lib/rust/parser/src/syntax/operator.rs]: We should remove all possible panics from the code. Maybe we should accept `NonEmptyVec` instead here and move the possible error handling to the call side?

File: lib/rust/parser/src/syntax/operator.rs [hunk 15]
```diff
@@ -45,77 +45,80 @@ impl<T> WithPrecedence<T> {
...
+        if tokens.len() < 2 {
             flattened.extend(tokens);
         } else {
-            let tokens = annotate_tokens_that_need_spacing(tokens);
+            let tokens = tokens.map(annotate_tokens_that_need_spacing);
             let ast = resolve_operator_precedence_internal(tokens);
             flattened.push(ast.into());
         }
     };
+    let breaks_group = |item: &syntax::item::Item| {
```
Comment (wdanilo) [inline lib/rust/parser/src/syntax/operator.rs]: what is "breaks_group"? This should have docs or more explicit name.

File: lib/rust/parser/src/syntax/operator.rs [hunk 16]
```diff
@@ -45,77 +45,80 @@ impl<T> WithPrecedence<T> {
...
+    process_no_space_group(&mut flattened, &mut no_space_group);
     resolve_operator_precedence_internal(flattened)
 }
 
-fn resolve_operator_precedence_internal(items: Vec<syntax::Item<'_>>) -> syntax::Tree<'_> {
+#[derive(PartialEq, Eq)]
+enum TokenType {
+    Ast,
+    Opr,
+}
```
Comment (wdanilo) [inline lib/rust/parser/src/syntax/operator.rs]: 1. This enum is kind of strange. We have `Token` defined and tokens cannot be AST.
2. Tokens has a lot of different types, none of them is AST.
3. This is defined in the "Precedence" section - is it a precedence-related definition?
4. It should have docs, as its usage patterns are not clear by looking at its name.

File: lib/rust/parser/src/syntax/operator.rs [hunk 17]
```diff
@@ -133,31 +136,30 @@ fn resolve_operator_precedence_internal(items: Vec<syntax::Item<'_>>) -> syntax:
                     // Prev operator in the [`operator_stack`] has a higher precedence.
                     let lhs = output.pop().map(|t| t.to_ast());
                     if lhs.is_none() { was_section_used = true; }
-                    let ast = syntax::Tree::opr_app(lhs, prev_opr.elem, Some(rhs.to_ast()));
+                    let ast = syntax::tree::apply_operator(lhs, prev_opr.elem, Some(rhs.to_ast()));
```
Comment (wdanilo) [inline lib/rust/parser/src/syntax/operator.rs:152]: Most of the items in tree are named as nouns, like "TypeDefinition". In fact, I've never seen AST that would use non-noun names. This should be named `operator_application` instead. (pls apply to other changed names).

File: lib/rust/parser/src/syntax/tree.rs [hunk 18]
```diff
@@ -194,11 +230,140 @@ impl<'s> span::Builder<'s> for Error {
 }
 
 
+// == Code Blocks ===
+
+/// A line of code.
+#[derive(Debug, Clone, PartialEq, Eq, Visitor, Reflect, Serialize, Deserialize)]
+pub struct Line<'s> {
```
Comment (wdanilo) [inline lib/rust/parser/src/syntax/tree.rs]: This should either be named `CodeBlockLine` or should be moved to `code_block` module. Right now, we have `syntax::tree::Line` and it doesnt make sense, as this is not "AST Line".

File: lib/rust/parser/src/syntax/tree.rs [hunk 19]
```diff
@@ -194,11 +230,140 @@ impl<'s> span::Builder<'s> for Error {
...
+    /// The content of the line, if any.
+    pub expression: Option<Tree<'s>>,
+    /// The end-of-line token.
+    pub newline:    token::Newline<'s>,
+}
```
Comment (wdanilo) [inline lib/rust/parser/src/syntax/tree.rs]: It's widely used standard that in parsers we remember the left spacing and left newline characters, not right ones. It would also make our life easier (e.g. consider a block that ends at the end of file - you dont have res enplane then).

File: lib/rust/parser/src/syntax/tree.rs [hunk 20]
```diff
@@ -194,11 +230,140 @@ impl<'s> span::Builder<'s> for Error {
...
+
+impl<'s> From<token::Newline<'s>> for Line<'s> {
+    fn from(newline: token::Newline<'s>) -> Self {
+        Self { expression: None, newline }
+    }
+}
+
+impl<'s> span::Builder<'s> for Line<'s> {
+    fn add_to_span(&mut self, span: Span<'s>) -> Span<'s> {
+        span.add(&mut self.newline).add(&mut self.expression)
```
Comment (wdanilo) [inline lib/rust/parser/src/syntax/tree.rs]: wait, so the newline is first? It is not in the type definition.

File: lib/rust/parser/src/syntax/tree.rs [hunk 21]
```diff
@@ -194,11 +230,140 @@ impl<'s> span::Builder<'s> for Error {
...
 pub type OperatorOrError<'s> = Result<token::Operator<'s>, MultipleOperatorError<'s>>;
 
+/// Return a reference to the first operator, whether that is the only operator, or the first of the
+/// `MultipleOperatorError` sequence.
+pub fn first_operator<'s, 'a>(opr: &'a OperatorOrError<'s>) -> &'a token::Operator<'s> {
+    match opr {
+        Ok(opr) => opr,
+        Err(oprs) => oprs.operators.first(),
+    }
+}
```
Comment (wdanilo) [inline lib/rust/parser/src/syntax/tree.rs]: why is it not a method? If you want to keep it as a global function, it needs better name, because now it is `tree:: first_operator` and its confusing. The same applies below.

File: lib/rust/parser/src/syntax/tree.rs [hunk 22]
```diff
@@ -231,6 +396,192 @@ impl<'s> span::Builder<'s> for MultiSegmentAppSegment<'s> {
...
+pub enum BlockBuilder<'s> {
+    /// The builder is in an indeterminate state until a non-empty line has been encountered, which
+    /// would distinguish an operator-block from a non-operator block.
+    Indeterminate {
+        /// The `Newline` token introducing the block, and `Newline` tokens for any empty lines
+        /// that have been encountered.
+        empty_lines: Vec<token::Newline<'s>>,
+    },
+    /// Building an operator block. If any line doesn't fit the operator-block syntax, that line
+    /// and all following will be placed in `body_lines`.
```
Comment (wdanilo) [inline lib/rust/parser/src/syntax/tree.rs]: these block-related things are big - can we refactor all of them to `tree::block` please?

File: lib/rust/parser/tests/parse.rs [hunk 23]
```diff
@@ -36,18 +37,49 @@ macro_rules! block {
...
+fn section_simple() {
+    let expected_lhs = block![(OprSectionBoundary (OprApp () (Ok "+") (Ident a)))];
+    test("+ a", expected_lhs);
+    let expected_rhs = block![(OprSectionBoundary (OprApp (Ident a) (Ok "+") ()))];
+    test("a +", expected_rhs);
 }
 
 #[test]
-fn type_definition_option() {
+#[ignore] // FIXME. This test catches a pre-existing bug.
```
Comment (wdanilo) [inline lib/rust/parser/tests/parse.rs]: 1. Fix left in the code
2. The fix does not explain anything.

File: lib/rust/parser/tests/parse.rs [hunk 24]
```diff
@@ -75,6 +107,72 @@ fn function_block_simple_args() {
...
+        "    +1",
+        "    bar",
+    ];
+    #[rustfmt::skip]
+    let expect = block![
+        (Assignment (Ident value) "="
+         (ArgumentBlockApplication (Ident foo) #(
+          (OprSectionBoundary (OprApp () (Ok "+") (Number 1)))
+          (Ident bar))))
+    ];
```
Comment (wdanilo) [inline lib/rust/parser/tests/parse.rs:175]: Please make sure that the tests cover all possible block types with all possible inonsistencies and errors.

**Review by wdanilo — CHANGES_REQUESTED** [review]

File: lib/rust/parser/src/lexer.rs (hunk 5 above)
Comment (kazcw) [reply lib/rust/parser/src/lexer.rs]: I understand and agree. I didn't write this :slightly_smiling_face:. However, this is not as bad as it looks: This panic is probably unreachable. It only panics if we try to end a block when there is no block started, and the following conditional ensures that cases like your example don't cause that to happen. I've added test cases for inconsistent indentation, and they pass. So, this is one of many panics that I intend to fix, but I don't think needs to be addressed before integration.

File: lib/rust/parser/src/syntax/operator.rs (hunk 14 above)
Comment (kazcw) [reply lib/rust/parser/src/syntax/operator.rs]: I've been thinking about doing just that to avoid this (and the `opt_rhs` panic below, and all the `if let` branches that should *always* match). However, the way the shunting yard algorithm shuffles values around, it's not trivial to prove to the typechecker the relevant property: *If the input is non empty, we will have at least one output*, which allows us to construct a return value. (If we don't prove it to the typechecker, we need to unwrap somewhere--or return an `Option`, but some callers would need to unwrap that). I have a way to do it in mind (and eliminate *all* the branches in this function that should never be taken), but I'll need to totally rewrite this function, so I've been planning on doing it later.

File: lib/rust/parser/src/syntax/operator.rs (hunk 16 above)
Comment (kazcw) [reply lib/rust/parser/src/syntax/operator.rs]: This replaces `last_token_was_ast`/`last_token_was_opr`. I have fixed the name--it actually classifies an `Item`, not a token. I have documented it and moved it into `resolve_operator_precedence_internal`, as it is only used to keep track of state in that function.

File: lib/rust/parser/src/syntax/tree.rs (hunk 19 above)
Comment (kazcw) [reply lib/rust/parser/src/syntax/tree.rs:242]: Either approach will work for us--in my design doc, I wrote that we can treat EOF as a zero-length newline, so that we handle files the same way whether or not they end in a newline character. We could instead insert a zero-length newline at the beginning of the file. However, that approach would be unconventional. It is common to treat a file that doesn't end with a newline as a special case. I have never encountered any format or application in which every line is expected to begin with a newline. Every editor I've ever used shows a file that starts with a newline as a file with a blank first line, because the universal standard is that a newline ends a line, rather than marking its start.

File: lib/rust/parser/src/syntax/tree.rs (hunk 20 above)
Comment (kazcw) [reply lib/rust/parser/src/syntax/tree.rs]: The inconsistency here was a mistake--I'm surprised to find the spans work out the same if these are in the wrong order, but looking at the implementation of `Span::concat_mut`, it happens to be commutative in some common cases (so there is probably no test case that would catch this mistake in this case).

File: lib/rust/metamodel/lexpr/src/lib.rs (hunk 1 above)
Comment (kazcw) [reply lib/rust/metamodel/lexpr/src/lib.rs:73]: I've elaborated on the docs of the method that provides the interface to this field, `ToSExpr::skip`

File: lib/rust/parser/src/syntax/operator.rs (hunk 17 above)
Comment (kazcw) [reply lib/rust/parser/src/syntax/operator.rs:152]: This is a helper function. It implements the rule that, in the RHS of an `OprApp`, a non-operator block is interpreted as a `BodyBlock` (whereas in other positions, it would be an argument block). Likewise, `apply` creates an `App`, `ArgumentBlockApplication`, or `OperatorBlockApplication`.

File: lib/rust/parser/src/syntax/item.rs (hunk 13 above)
Comment (kazcw) [reply lib/rust/parser/src/syntax/item.rs]: I would like to accept an `Item::Block`, but that's a variant, not a type, so the options are: Accept an `Item`, and then deal with a non-`Block` branch that should never happen, or accept the contents of `Item::Block`. Here I've done the latter.

File: lib/rust/parser/tests/parse.rs (hunk 24 above)
Comment (kazcw) [reply lib/rust/parser/tests/parse.rs:175]: I added a lot of weird-indentation test cases; they all parse correctly, but in many cases the whitespace is printed back wrong. I'm looking in to that.

**Comment by kazcw** [comment]
I've merged #3591 into this, because that depends on a bugfix I've implemented here.

File: lib/rust/parser/tests/parse.rs (hunk 24 above)
Comment (kazcw) [reply lib/rust/parser/tests/parse.rs:175]: All the whitespace problems were caused by an incorrect implementation of `span::Builder for Option`.

File: lib/rust/parser/tests/parse.rs (hunk 23 above)
Comment (kazcw) [reply lib/rust/parser/tests/parse.rs]: Fixed it (see below).

File: lib/rust/prelude/src/data/empty_vec.rs (hunk 4 above)
Comment (kazcw) [reply lib/rust/prelude/src/data/empty_vec.rs]: I renamed it to `VecAllocation`, which I think is clearer, and added an example.

File: lib/rust/parser/src/lexer.rs (hunk 2 above)
Comment (kazcw) [reply lib/rust/parser/src/lexer.rs]: Renamed it, documented it.

File: lib/rust/parser/src/lexer.rs (hunk 5 above)
Comment (kazcw) [reply lib/rust/parser/src/lexer.rs]: I realized this panic is easily avoidable, so I went ahead and fixed it.

File: lib/rust/parser/src/macros/resolver.rs (hunk 11 above)
Comment (kazcw) [reply lib/rust/parser/src/macros/resolver.rs]: The types in this library have me questioning my love for Rust's pattern matching

**Comment by kazcw** [comment]
Besides the changes discussed in the individual review comments, I've: implemented the non-recursive `resolver::run` I mentioned earlier; wrote a benchmark that parses a (deterministically) generated large program with various depths of nested blocks, and covers all the block types and the expression-to-statement transform. With benchmark in hand, I reduced memory allocation a bit, improving performance on the benchmark 5x.

File: lib/rust/parser/src/syntax/operator.rs (hunk 14 above)
Comment (wdanilo) [reply lib/rust/parser/src/syntax/operator.rs]: Why don't we accept NonEmptyVec and we convert it to vec just in the first line and then we will leave the panic in the code with comment that its unreachable because there is always at least one arg?

File: lib/rust/parser/src/lib.rs [hunk 25]
```diff
@@ -262,4 +225,44 @@ mod benches {
...
+        let mut indent = 0u32;
+        for _ in 0..lines {
+            // Indent:
+            // 1/8 chance of increasing.
+            // 1/8 chance of decreasing.
+            // 3/4 chance of leaving unchanged.
+            match rng.gen_range(0..8) {
+                0u32 => indent = indent.saturating_sub(1),
+                1 => indent += 1,
+                _ => (),
```
Comment (wdanilo) [inline lib/rust/parser/src/lib.rs:245]: Are we sure that we want to use random input in benchmarsk? This way it would be hard to compare benches between different commits.

File: lib/rust/parser/src/macros/resolver.rs [hunk 26]
```diff
@@ -186,43 +195,61 @@ enum Step<'s> {
...
+    fn replace_current_with_parent_macro(&mut self, parent_macro: PartiallyMatchedMacro<'s>) {
+        let child_macro = mem::replace(&mut self.current_macro, parent_macro);
+        self.current_macro.push(child_macro);
     }
 
-    fn replace_current_with_parent_macro(&mut self, mut parent_macro: PartiallyMatchedMacro<'s>) {
-        mem::swap(&mut parent_macro, &mut self.current_macro);
-        let child_macro = parent_macro;
-        self.current_macro.current_segment.body.push(child_macro.into());
+    fn macro_scope_start(&self) -> usize {
```
Comment (wdanilo) [inline lib/rust/parser/src/macros/resolver.rs:254]: This needs docs.

File: lib/rust/parser/src/macros/resolver.rs [hunk 27]
```diff
@@ -186,43 +195,61 @@ enum Step<'s> {
     MacroStackPop(syntax::Item<'s>),
 }
 
+/// Information about macro resolution state that is stored while processing a deeper indentation
+/// level.
+#[derive(Debug)]
+struct Scope<'s> {
+    parent_tokens: std::vec::IntoIter<syntax::Item<'s>>,
+    macros_start:  usize,
+    outputs_start: usize,
```
Comment (wdanilo) [inline lib/rust/parser/src/macros/resolver.rs:228]: I dont understand what these numbers are. I've read the docs, but its still not clear to me. I mean, I understand it, but its hard to understand it without our domain knowledge. Could you please add docs to Scope or to Resolver (even better) explaining how nested macros are resolved in case of blocks, please? It's worth explaining that blocks have precedence over macros. For example, 

```
if foo
    then bar
```

would not parse as "if ... then ..." - it will parse as "if foo" (not fully matched macro) applied to a code block. It's important to describe that in the docs - either here or in the global docs in the parser module.

File: lib/rust/parser/src/macros/resolver.rs [hunk 28]
```diff
@@ -247,20 +326,54 @@ impl<'s> Resolver<'s> {
...
                 }
             }
         }
+        syntax::tree::block::body_from_lines(self.lines)
+    }
 
-        event!(TRACE, "Finishing resolution. Popping the macro stack.");
-        while let Some(parent_macro) = self.macro_stack.pop() {
-            self.replace_current_with_parent_macro(parent_macro);
+    fn exit_scope(&mut self, scope: Scope<'s>) -> std::vec::IntoIter<syntax::Item<'s>> {
```
Comment (wdanilo) [inline lib/rust/parser/src/macros/resolver.rs]: Add docs here please. It's clear what this function does based on its name, but its so long and complex, that a short docs explaining corner cases or interesting properties will be very appreciated. If you dont feel that such docs make sense or there are no really interesting corner cases covered by this function, feel free to skipping this comment :)

File: lib/rust/parser/src/syntax/item.rs [hunk 29]
```diff
@@ -73,6 +83,34 @@ impl<'s> TryAsRef<Item<'s>> for Item<'s> {
     }
 }
 
+/// Given a sequence of `Item`s belonging to one block, create an AST block node, of a type
```
Comment (wdanilo) [inline lib/rust/parser/src/syntax/item.rs]: You can refer to ``` [`Item`] ``` in the docs - we try to use this syntax whenever its possible to refer to exisitng entities.

**Review by wdanilo — APPROVED** [review]

File: lib/rust/parser/src/lib.rs (hunk 25 above)
Comment (kazcw) [reply lib/rust/parser/src/lib.rs:245]: `ChaCha8` is deterministic; it produces different output streams for different seeds, but I've initialized it with a fixed seed (0).

**Review by wdanilo — APPROVED** [review]
